### PR TITLE
Add missing tiles to arrakis.yaml and fixed the map import

### LINKED ISSUE #12202: D2k arrakis.yaml tilesets mixed up
Bug description and Fix suggestion.

In an attempt to use OpenRA.Utilities.exe to import mission A1V1 i noticed some misplaced tiles.
Using the correct tileset for this map `BLOXTREE.R8` (phrase "BLOXTREE" found in a original D2k .Sav file of the mission) only some of the misplaced tiles were correct now.

I compared in the map editor with original game "screenshots" and found:
Rock-Detail 94 should be Rotten-Base 251
Rotten Base 250 and 251 together should be Sand-Detail 117

In `arrakis.yaml` (tileset file of d2k) i found the corresponding entries under `ID:94, ID:250, ID:251`
Replacing`Images: BLOXBASE.R8` with `Images: BLOXTREE.R8` fixed the issue. The map is now displayed correctly with no manual edit.
capture

Is this the right way to do it? I guess one could find the correct frames of  `ID:94 ID:250 ID:251` as an alternative way to fix it.
Maybe someone with insight to the underlying mechanics can deduce a general fix for all tiles/tilesets?

One thing i noticed after the fix: the tiles like       

```
Tiles:
            0: Sand
```

show up in wrong color in the map preview. i think the tiletype would have to be corrected as well...

Cheers,
    SirCake

### PR DESCRIPTION
**Pull request by abcdefg30**
My take on #12244.
Fixes #12202.

### CODE REVIEW

**Comment by CH4Code** [comment]
All right, finaly someone with the "right" approach :)

**Comment by ltem** [comment]
I converted all the multiplayer maps and checking them right now.

The only error I found so far is one missing/wrong tile.
Map: 2PLAY13
Tileset: Bloxwast.r8

| Original | OpenRA|
|:-------:|:-------:|
|image | image |

The tile id from the right picture is 65 but I didn't found the correct tile in OpenRA.

**Comment by ltem** [comment]
So I finally tested the mission maps now and I found only two tile errors.

Map: O7V1
Tileset: bloxwast

| Original | OpenRA |
|:------:|:------:|
| image | image |
| | conclusion |

-------------------------------

__Additional notes:__
I don't know if the following points are in scope of the importer but I stumbled over them during the review process
- __Note 1:__ Some buildings (Corrino?) won't be imported correctly.

| Original | OpenRA |
|:------:|:------:|
| image | image  |
| image | image |


- __Note 2:__ If a map has some kind of "padding" the importer will crash


|scali|
|:------------:|
|Map with black padding|

```
$ ./OpenRA.Utility.exe d2k --import-d2k-map missionMaps/A2V2.MAP bloxtree.r8
System.IndexOutOfRangeException: Array index is out of range.
  at OpenRA.CellLayer`1[OpenRA.TerrainTile].set_Item (CPos cell, TerrainTile value) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.D2k.UtilityCommands.D2kMapImporter.FillMap () [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.D2k.UtilityCommands.D2kMapImporter..ctor (System.String filename, System.String tileset, OpenRA.Ruleset rules) [0x00000] in <filename unknown>:0
```

@pchote mentioned that a fix for such maps, could cause bogus imports for others. 
This only effects the first 4 Missions of every faction (A1V1, A1V2, A2V1,....), which are already imported by @CH4Code (https://github.com/CH4Code/D2k-Campaign-Resources), so ¯\\\_(ツ)\_/¯

**Comment by MustaphaTR** [comment]
Side specific structure images in OpenRA are set according to side itself. If side is Harkonnen and not Corrino it is normal that it use Harkonnen Palace. Importer doesn't create players afaik, unless this PR added so.

Corrino, Mercenaries, Fremen and Smugglers are valid sides on OpenRA.

**Comment by CH4Code** [comment]
I originaly didn't check the multiplayer maps for tile errors, since those were allready (manually) fixed and done (in the official map pool). So abcdefg30's corrected importer based on my (not meant to be overarching) fixes should only work 100% in case of missions.
@Item: The 1st 4 missions in my repo should not **all** display correctly in vanilla OpenRA, since they use a custom tileset (arrakis.yaml). Whould be surprised if they did.

**Comment by CH4Code** [comment]
I could extend the custom tileset, which abcdefg30 then could translate into c#/map importer, fixing all multiplayer maps. But I will **only** do this if it is requested by abcdefg30 or pchote. It's a ton of work.

**Comment by MustaphaTR** [comment]
Those black area in maps added on Gruntmods edition and doesn't exist in either original 1.2 or 1.6. It is added for Hi-Res Patch or they was crashing.

**Comment by abcdefg30** [comment]
@ltem thanks for going through all the missions again.

So as far as I understand, this only seems to fix the tileset for missions.
What about taking this (after I fixed O7V1) and adding the required new tiles as needed when importing the multiplayer maps? (So this PR doesn't get any larger.)

File: mods/d2k/tilesets/arrakis.yaml [hunk 1]
```diff
@@ -5671,3 +5671,381 @@ Templates:
...
+		Id: 10298
+		Images: BLOXBGBS.R8
+		Frames: 707, 708
+		Size: 2,1
+		Category: Rock-Detail
+		Tiles:
+			0: Rough
+			1: Rough
+	Template@1084:
+		Id: 10309
```
Comment (MustaphaTR) [inline mods/d2k/tilesets/arrakis.yaml]: This and 4 above don't have matching @ part and Id:

**Comment by CH4Code** [comment]
@abcdefg30 seems fine, #12788 is for multiplayer maps and could have it's own pull request. Having the converter fixed for Missions will greatly benefit the campaign ticket (#9287).

**Comment by ltem** [comment]
@abcdefg30 I'm totally fine with that

Most of the multiplayer tile errors seem to be caused by bloxxmas, which wasn't used for the missions, so it seems fitting to fix this with an other PR.

File: mods/d2k/tilesets/arrakis.yaml (hunk 1 above)
Comment (abcdefg30) [inline mods/d2k/tilesets/arrakis.yaml]: Ops, I forgot to rename those. Thanks.

**Comment by abcdefg30** [comment]
Updated and fixed the remaining two tiles.

**Comment by ltem** [comment]
O7V1 is fixed so :+1: from me

**Review by ltem — APPROVED** [review]

**Comment by CH4Code** [comment]
Hi, went through the bugs in the MP maps as well. Those additional tiles (no duplicates) will fix importing all MP maps. 
Also I added a "Bridge" Category for all the bridge tiles. On 4P10, 6P4, 6P5 however, the bridges are partially only passable by infantry(due to other predefined tiles) - but manual fix in like 1 minute with the new tiles.
@abcdefg30 just add "Bridge" category and add these lines to your arrrakis.yaml:
[http://pastebin.com/C6xzM4xZ](http://pastebin.com/C6xzM4xZ) (valid for one week)
This will finalize our map-import-effort I hope :)

**Comment by abcdefg30** [comment]
Thanks. Added it.

File: mods/d2k/tilesets/arrakis.yaml [hunk 2]
```diff
@@ -6049,3 +6049,236 @@ Templates:
 		Tiles:
 			0: Rough
 			1: Rough
+	Template@1085:
+		Id: 1085
+		Images: BLOXXMAS.R8
+		Frames: 345, 346
+		Size: 2,1
+		Category: Bridge
```
Comment (reaperrr) [inline mods/d2k/tilesets/arrakis.yaml]: You need to add this category to `EditorTemplateOrder` to make them available in the editor (I suggest placing it before `Unidentified`).

**Comment by reaperrr** [comment]
In addition to the category fix, the destroyed north/south bridge ends are currently missing.

I've created yaml definitions here:
https://gist.github.com/reaperrr/054c08bb66e6cbd8fda889e3d434d521

@Item @CH4Code can you confirm that they look as they should and have correct (im)passability?

**Comment by CH4Code** [comment]
> In addition to the category fix, the destroyed north/south bridge ends are currently missing.

Yes and no. Look at Template 513 and 514 in rock detail (where the rest of the bridges are). If you copy/paste parts of it you can construct the big destroyed north/south ends. Thats why the map is generated correctly without them "missing", as is. 

The openRA devs originaly simply misinterpreted the BLOX* tile file, i guess (which is't hard at all :D ). There, the bridges are dispalyed in a "compressed" format.

> can you confirm that they look as they should and have correct (im)passability?

They are nor passable at all they way you defined them. Judging from  513 and 514 the centermost bridge tile should be "Rough"(passable for inf), the other "Cliff". 

@reaperrr I don't think it is needed to fix importing, though.
Of course detatched ends have better useability in map editor :)

**Comment by ltem** [comment]
Uhh, I derped a bit, I should have seen that.

In the original tileset, the bridges are smaller.
image
Source: http://old.d2kplus.com/wiki/index.php?title=File:Bloxxmas.png


Infantry can pass the bridge "slopes" (but not vehicles), the cursor gets grey so it should be `Rough`
image
image


I would move the tiles 513, 514, 515 as well to the `Bridge` category.
image

**Comment by CH4Code** [comment]
@Item I think those are map editor issues and should get their own PR later.

**Comment by CH4Code** [comment]
> Yeah "compressed" sounds very good. I don't know what is to prefer for map makers.

Smaller, but identifyable "building blocks" imho. But, if we remove/reorganize existing tiles alltogether, that surely will cause problems for existing maps.

**Comment by CH4Code** [comment]
Nice :) Found the last fix :) Replace 
513    4: Rough -> 4: Transition    and  
514    4: Rough -> 4: Transition.
Then all passability issues are auto-fixed :)

**Comment by ltem** [comment]
I agree with you  about the smaller building blocks and as you already said before (^^ I tried to :+1: it up but it was already gone) this should be done in an extra PR.

**Comment by abcdefg30** [comment]
Updated. Thanks @reaperrr (and @CH4Code and @ltem for constantly rechecking things ^^).

**Comment by CH4Code** [comment]
@abcdefg30: you'll need to incorporate my comments on 513, 514, 515 and my comment on reaperrs added tiles, too, or well get passability issues needing post-map-generation manual fixes.

File: mods/d2k/tilesets/arrakis.yaml [hunk 3]
```diff
@@ -5671,3 +5671,650 @@ Templates:
...
+			1: Transition
+	Template@1109:
+		Id: 1109
+		Images: BLOXXMAS.R8
+		Frames: 347, 348
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
+			1: Cliff
```
Comment (CH4Code) [inline mods/d2k/tilesets/arrakis.yaml]: 1: Rough

File: mods/d2k/tilesets/arrakis.yaml [hunk 4]
```diff
@@ -5671,3 +5671,650 @@ Templates:
...
+	Template@1111:
+		Id: 1111
+		Images: BLOXXMAS.R8
+		Frames: 390, 391
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
```
Comment (CH4Code) [inline mods/d2k/tilesets/arrakis.yaml]: 1: Rough

File: mods/d2k/tilesets/arrakis.yaml [hunk 5]
```diff
@@ -5671,3 +5671,650 @@ Templates:
...
+			1: Cliff
+	Template@1110:
+		Id: 1110
+		Images: BLOXXMAS.R8
+		Frames: 348, 349
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
```
Comment (CH4Code) [inline mods/d2k/tilesets/arrakis.yaml]: 0: Rough

File: mods/d2k/tilesets/arrakis.yaml [hunk 6]
```diff
@@ -5671,3 +5671,650 @@ Templates:
...
+			1: Cliff
+	Template@1112:
+		Id: 1112
+		Images: BLOXXMAS.R8
+		Frames: 391, 392
+		Size: 2,1
+		Category: Bridge
+		Tiles:
+			0: Cliff
```
Comment (CH4Code) [inline mods/d2k/tilesets/arrakis.yaml]: 0: Rough

File: mods/d2k/tilesets/arrakis.yaml [hunk 7]
```diff
@@ -5671,3 +5671,650 @@ Templates:
 		Category: Sand-Detail
 		Tiles:
 			0: Rough
```
Comment (CH4Code) [inline mods/d2k/tilesets/arrakis.yaml]: @ Id: 513
Tiles:
     4: Transition

@ Id: 514
Tiles:
    4: Transition

@ Id: 515
Tiles:
			1: Cliff
			2: Cliff

**Review by CH4Code — CHANGES_REQUESTED** [review]
Change "Tiles:" to fix remaining passability issues.

**Comment by CH4Code** [comment]
Also ..... just some cosmetics : could you move` Template@1108:` to be before `Template@1085: `?
Has nice (originally intended, messed up ^^) effect on tile order in editor:
grafik

**Comment by abcdefg30** [comment]
Sorry for the delay, updated.

**Comment by ltem** [comment]
The end is near :D

I went through all the multiplayer maps and the only tile error if found is Error 3 of my comment https://github.com/OpenRA/OpenRA/pull/12784#issuecomment-280706706. It still exits (Map: [6PLAY3.MAP.zip](https://github.com/OpenRA/OpenRA/files/819406/6PLAY3.MAP.zip) Tileset: bloxwast.r8)

So after Error 3 is fixed my :+1: valid again

**Comment by CH4Code** [comment]
Yep, those two errors are C# solvable only.  👍, too, after this :)

**Comment by abcdefg30** [comment]
Sorry for the late update. Finally fixed it. This should be good to go now.

**Comment by CH4Code** [comment]
Erm, can't make out any changes since last time o.O

**Review by ltem — APPROVED** [review]
6PLAY3 fixed

**Review by CH4Code — APPROVED** [review]
That last change finally fixed all optical conversion problems. *thumbsup*

**Comment by abcdefg30** [comment]
Rebased again.

> Erm, can't make out any changes since last time o.O

I squashed the changes into existing commits. I just added https://github.com/OpenRA/OpenRA/pull/12784/commits/80f4feeab30648148e43de75a26bf3d620576781#diff-54b547985b15dbd63a5da7db1eb43a54R422 (the tile 579 is part of the template 101) and https://github.com/OpenRA/OpenRA/pull/12784/commits/80f4feeab30648148e43de75a26bf3d620576781#diff-54b547985b15dbd63a5da7db1eb43a54R413 (tile 342 is part of template 220).

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d4d278844878acd3ee92da7a9c5665e0ab263ec2...9bc45e9c364cf7fea6c17c43071e29cbaa5d1ab8)

**Pull request merged**